OpenSim/Common/MultivariatePolynomialFunction.cpp

```diff
@@ -17,205 +17,362 @@
+                "polynomial dimension ({}), but it does not.", 
+                vars.size(), dimension)
+
+        std::vector<std::vector<int>> combinations;
+        generateCombinations(dimension, order, combinations);
+        for (int i = 0; i < coefficients.size(); ++i) {
+            Term term;
+            for (int j = 0; j < dimension; ++j) {
+                if (combinations[i][j] > 0) {
+                    term[vars[j]] = combinations[i][j];
+                }
+            }
+            terms.at(i) = term;
+        }
+        
+        for (int i = 0; i < coefficients.size(); ++i) {
+            (*this)[terms[i]] = coefficients[i];
+        }
+    }
+
+    // CALC METHODS
+    /**
+     * Calculate the order of the polynomial.
+     */
+    int calcOrder() const {
+        int order = 0;
+        for (auto it = begin(); it != end(); it++) {
+            int sum = 0;
+            for (auto it2 = it->first.begin(); it2 != it->first.end(); it2++) {
+                sum += it2->second;
+            }
+            if (sum > order) {
+                order = sum;
+            }
+        }
+        return order;
+    }
+
+    /**
+     * Calculate the vector of coefficients of the polynomial for a given order
+     * and dimension. 
+     */
+    SimTK::Vector calcCoefficients(int dimension, int order, 
+            const std::vector<std::string>& vars) const {
+
+        OPENSIM_THROW_IF(static_cast<int>(vars.size()) != dimension, 
+                Exception,
+                "Expected the number of variable names ({}) to match the "
+                "polynomial dimension ({}), but it does not.", 
+                vars.size(), dimension)
+        
+        const MultivariatePolynomial& poly = *this;
+        MultivariatePolynomial temp(dimension, order, vars);
+        int numCoefficients = choose(dimension + order, order);
+        SimTK::Vector coefficients(numCoefficients, 0.0);
+        for (auto it2 = temp.begin(); it2 != temp.end(); ++it2) {
+            auto it = poly.find(it2->first);
+            if (it != poly.end()) {
+                coefficients[it2->second] = it->second;
+            }
+        }
+
+        return coefficients;
+    }
+
+    // POLYNOMIAL OPERATIONS
+    /**
+     * Get a new polynomial that is the first derivative of the current 
+     * polynomial with respect to the specified variable. 
+     */
+    MultivariatePolynomial getDerivative(const std::string& var) const {
+        const MultivariatePolynomial& poly = *this;
+        MultivariatePolynomial deriv;
+        for (auto it = poly.begin(); it != poly.end(); it++) {
+            Term t = it->first;
+            double c = it->second;
+            for (auto it2 = t.begin(); it2 != t.end(); it2++) {
+                Term t2 = t;
+                if (it2->first == var) {
+                    t2.erase(it2->first);
+                    if (it2->second > 1) {
+                        t2[it2->first] = it2->second - 1;
```

Comment:
It would be really great to add an accessing member function that clarifies what we get, e.g.:
`double getCoefficient(const PowerProduct& key)`
and similarly for the map of variable to power.
Because we have two nested maps, i really have to read carefully to understand which one i am accessing, the coefficient, or the power.